### PR TITLE
[#121483833] Capture API calls during pipeline tests

### PR DESCRIPTION
**Pull request by henrytk**
## What

[#121483833 - Capture api calls made in pipeline tests](https://www.pivotaltracker.com/story/show/121483833)

We want to be able to capture the CF_TRACE output during Cloud Foundry tests in the pipeline, because it provides useful debugging information when investigating failed tests. We do not want to output this to the pipeline, as we cannot be certain it is free of secrets, so we will store it in a bucket.

## How to review

* Deploy from this branch
* Check the `upload-test-artifacts` task succeeds for the following jobs:
    * acceptance-tests
    * custom-acceptance-tests
    * smoke-tests
* Check the bucket `gds-paas-$DEPLOY_ENV-test-artifacts` for three files, which should be TAR files from the jobs listed above.
* Make one of the tests fail and check it still uploads the file.
* If you are really curious you can download them, un-TAR them and look inside.

## Who can review

Anyone but me.

### CODE REVIEW

**Comment by henrytk** [comment]
The Travis build is currently failing due to an unrelated issue with Go dependencies.

**Comment by saliceti** [comment]
I'm not sure the `s3init` refactor is necessary. "s3init" has a special meaning emphasizing that a file is uploaded only once. You don't need this logic here and could do simply `aws s3 cp`.

File: concourse/tasks/smoke-tests-run.yml [hunk 1]
```diff
@@ -24,3 +26,8 @@ run:
       ln -snf $(pwd)/cf-release/src/smoke-tests /var/vcap/packages/smoke-tests/src/github.com/cloudfoundry/cf-smoke-tests
       set -o pipefail
       /var/vcap/jobs/smoke-tests/bin/run
+
```
Comment (saliceti) [inline concourse/tasks/smoke-tests-run.yml]: This block will not be executed if the smoke test fails. Can we do the upload in an `ensure` block?

**Comment by henrytk** [comment]
I have removed the refactoring of the s3init.sh script and used AWS CLI instead. I have also wrapped the upload task in an ensure block.

File: concourse/tasks/upload-test-artifacts.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+run:
+  path: sh
+  args:
+    - -e
+    - -c
+    - |
+      echo "Archiving test artifacts..."
+      RAND=$(hexdump -n 2 -e '/2 "%u"' /dev/urandom)
+      ARTIFACT_NAME="test-artifact-$(date '+%Y-%m-%d-%H-%M-%S')-$RAND.tgz"
+      tar -C artifacts -czvf "${ARTIFACT_NAME}" ./
```
Comment (saliceti) [inline concourse/tasks/upload-test-artifacts.yml]: It looks like tar doesn't like "." :

```
$ tar xvzf test-artifact-2016-11-03-12-19-56-17410.tgz
x ./: Can't update time for .
x ./acceptance_tests/
x ./acceptance_tests/CATS-TRACE-CATS-1.txt.log
x ./acceptance_tests/CATS-TRACE-CATS-2.txt.log
x ./acceptance_tests.log
x ./acceptance_tests.err.log
tar: Error exit delayed from previous errors.
```

I would suggest something more explicit such as:

```
(cd artifacts && tar -czvf "${ARTIFACT_NAME}" *)
```

**Pull request merged**